# 📊 Customer Churn Prediction

## 📌 Project Introduction

Customer churn is one of the most important business challenges for companies that rely on long-term customer relationships.

After analyzing customer behavior in the previous Customer Churn Analysis project, the next logical step is to move from descriptive analytics to predictive analytics.

In this notebook, an end-to-end Machine Learning workflow is developed to estimate the probability of customer churn. The project covers data preprocessing, feature engineering, model training, evaluation, and business interpretation, demonstrating how predictive models can support customer retention strategies and data-driven decision making.

## 🛠 Workflow

```text
Read Dataset
      ↓
Data Cleaning
      ↓
Null Values
      ↓
Encoding
      ↓
Feature Selection
      ↓
Train/Test Split
      ↓
Model Training
      ↓
Evaluation
      ↓
Business Interpretation
```

In [ ]:
#Imports
import pandas as pd
import numpy as np


In [ ]:
#machine learning
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
)

In [ ]:
# visualization
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# Dataset
df= pd.read_csv('telco_customer_churn.csv', sep=';')

# Read the dataset

In [ ]:
#First Approach
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# Check the variables of interest
## Churn,MonthlyCharges,TotalCharges

In [ ]:
print(df['Churn'].dtype)
print(df['MonthlyCharges'].dtype)
print(df['TotalCharges'].dtype)

object
float64
object


# I correct the error (TotalCharges is changed to float type)

In [ ]:
df['TotalCharges']= pd.to_numeric(df['TotalCharges'],errors='coerce')

print(df['TotalCharges'].dtype)
print(df['TotalCharges'].isnull().sum())

float64
11


# Null Check

In [ ]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


# Drop the NULLS

In [ ]:
df=df.dropna(subset=['TotalCharges']) 
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


# Convert Churn to Binary

In [ ]:
df['Churn'].value_counts()

,count
Churn,
No,5163
Yes,1869


In [ ]:
#I map and convert "Yes" and "No" to 1 and 0.
df['Churn']=df['Churn'].map({'Yes':1,'No':0})
df['Churn'].value_counts()

,count
Churn,
0,5163
1,1869


## Why I'm doing this: The model needs numbers.

### It doesn't understand "Yes" or "No"; it only understands numbers.

## What it means: <br>
### 1 = customer leaves <br>
### 0 = customer stays

# Churn Average Ckeck

In [ ]:
print(df['Churn'].mean())

0.26578498293515357


# The company has a customer churn rate of 26%.

# Choosing features

In [ ]:
features = [
    'tenure',
    'MonthlyCharges',
    'TotalCharges',
    'Contract',
    'InternetService',
    'OnlineSecurity',
    'PaymentMethod'
]
X=df[features]
y=df['Churn']

# convert categorical features (ONE HOT)

In [ ]:
X= pd.get_dummies(X,drop_first=True) # This transforms text into numbers that the model understands.

# Train/Test Split

In [ ]:
from sklearn.model_selection import train_test_split

X_train,X_test, y_train,y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# I have just split the data into:

## 80% training
## 20% testing

# To validate whether the model generalizes.

# Train the Model

In [ ]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

# Model Evaluation

In [ ]:
from sklearn.metrics import classification_report, accuracy_score

y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.7896233120113717
              precision    recall  f1-score   support

           0       0.84      0.89      0.86      1033
           1       0.62      0.52      0.57       374

    accuracy                           0.79      1407
   macro avg       0.73      0.70      0.71      1407
weighted avg       0.78      0.79      0.78      1407



# Comparison V1 vs V2
 Before (numerical features only)

*   Recall churn → 0.43
*   F1 churn → 0.51

  Now (with business features)

*   Recall churn → 0.52 ✅
*   F1 churn → 0.57 ✅
	​
<br>


# **This indicates that 20% more at-risk customers are being detected**


# **Project Summary**

### By adding categorical variables significantly increased the model's ability to detect customers at risk of churn.

### The model improved with:

* Contract
* InternetService
* OnlineSecurity
* PaymentMethod

### That confirms something very important:

# Churn depends not only on price, but also on the type of relationship with the customer.

## What it means for business

### Before:

* Many customers were slipping away.

### Now:

* More customers are identified in time.

*   Action can be taken sooner (retention).

### 📚 Key Learnings

Throughout this project I developed an end-to-end machine learning workflow for customer churn prediction.

Key concepts practiced:

✔ Data Cleaning

✔ Feature Engineering

✔ One-Hot Encoding

✔ Train/Test Split

✔ Logistic Regression

✔ Classification Metrics

✔ Business Interpretation

This project demonstrates how predictive analytics can transform historical customer data into actionable insights for retention strategies.